clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -392,44 +396,57 @@ private void throwIfRecordTooLarge() throws RecordTooLargeException {
             throwIfUnauthorizedTopics();
             throwIfRecordTooLarge();
 
-            for (PartitionRecords<K, V> part : this.records) {
-                if (!subscriptions.isAssigned(part.partition)) {
-                    // this can happen when a rebalance happened before fetched records are returned to the consumer's poll call
-                    log.debug("Not returning fetched records for partition {} since it is no longer assigned", part.partition);
-                    continue;
-                }
-
-                // note that the consumed position should always be available
-                // as long as the partition is still assigned
-                long position = subscriptions.position(part.partition);
-                if (!subscriptions.isFetchable(part.partition)) {
-                    // this can happen when a partition is paused before fetched records are returned to the consumer's poll call
-                    log.debug("Not returning fetched records for assigned partition {} since it is no longer fetchable", part.partition);
-                } else if (part.fetchOffset == position) {
-                    long nextOffset = part.records.get(part.records.size() - 1).offset() + 1;
-
-                    log.trace("Returning fetched records at offset {} for assigned partition {} and update " +
-                            "position to {}", position, part.partition, nextOffset);
-
-                    List<ConsumerRecord<K, V>> records = drained.get(part.partition);
-                    if (records == null) {
-                        records = part.records;
-                        drained.put(part.partition, records);
-                    } else {
-                        records.addAll(part.records);
-                    }
+            int maxRecords = maxPollRecords < 0 ? Integer.MAX_VALUE : maxPollRecords;
```

Comment:
Could then add an `atLeast` validator too. What should happen in the user sets 0? should it be `atLeast(1)`?